### DISCUSSION TITLE
Hello from LeMP-repl

### DISCUSSION THREAD

**Discussion by jonathanvdc** [qwertie/ecsharp#55] [General]
Hi there @qwertie! Long time no see.

Just wanted to drop by and tell you that I created [`LeMP-repl`](https://github.com/jonathanvdc/LeMP-repl). It's a simple interactive program that takes EC# code as input and prints out equivalent macro-expanded LES code. It's a useful tool for people who want to develop things based on EC#, so I thought I'd tell you about it. Might even be worth a mention in a wiki page or something.

Well, that's about it I guess. How are you doing?

**Comment by qwertie** [comment]
I just want to make sure you knew about `LeMP.exe --editor` because that's what I use for this purpose. By changing `--outext=out.cs` to `--outext=out.les` the editor produces LES output. I guess LeMP-repl is more convenient for those who like the command line. 

Thanks for stopping by. I've been having, let's say, a mid-life crisis after I flew to North America for 3 job interviews and wasn't hired. I've been meaning to update EC# to support C# 7 and 7.2 syntax, which really isn't hard to do (EC# already supported some of the things like inner functions, after all, so all I needed to add for C# 7 was the `x is Foo y` pattern matching, and tuples-as-types, which is also not difficult thanks to EC#'s three-phase parsing system)... but I haven't got around to it simply due to lack of motivation. If you tell me you want these features, boom, I'll have them done pronto.

It seems to me like a simple REPL like this should be part of LeMP.exe.

You know what we really need? A way to run any arbitrary code at compile time - and share code between compile-time and run-time - so that LeMP will be clearly superior to T4 templates. I know you wanted to do that by writing a whole EC# compiler. But I would happily settle for invoking Roslyn to do the job.

**Comment by jonathanvdc** [comment]
> I just want to make sure you knew about `LeMP.exe --editor` because that's what I use for this purpose. By changing `--outext=out.cs` to `--outext=out.les` the editor produces LES output.

I often feel like you've already built every conceivable gizmo on top of Loyc and I'm just seeing the tip of the iceberg. :) Will use `LeMP.exe --editor` in the future.

> Thanks for stopping by. I've been having, let's say, a mid-life crisis after I flew to North America for 3 job interviews and wasn't hired.

I'm so sorry to hear that. :(

> I've been meaning to update EC# to support C# 7 and 7.2 syntax, which really isn't hard to do (EC# already supported some of the things like inner functions, after all, so all I needed to add for C# 7 was the `x is Foo y` pattern matching, and tuples-as-types, which is also not difficult thanks to EC#'s three-phase parsing system)... but I haven't got around to it simply due to lack of motivation. If you tell me you want these features, boom, I'll have them done pronto.

I'm currently doing complete rewrite of Flame from the bottom up. I'm really happy with how [Pixie](https://github.com/jonathanvdc/Pixie) turned out. I've also started working on a (C#) rewrite of the core Flame framework on [a separate branch](https://github.com/jonathanvdc/flame/tree/csharp-rewrite). When I'm done with that, I think I'll tweak `ecsc`'s codegen functionality to target the new and improved Flame. At *that* point, C# 7 support would be most welcome.

The rewrite isn't proceeding as fast as I'd hope. Mostly because of uni group projects&mdash;I have like five of them right now and they're consuming nearly all of my time during the semester. Writing code for all of those projects also kind of saps me of the energy to write code for my own projects. Hopefully I'll have some time to work on those over the summer. I'll have my full Flame/`ecsc` rewrite *eventually.*

> You know what we really need? A way to run any arbitrary code at compile time - and share code between compile-time and run-time - so that LeMP will be clearly superior to T4 templates. I know you wanted to do that by writing a whole EC# compiler. But I would happily settle for invoking Roslyn to do the job.

That'd be really nice. But maybe we should also try to make EC# more accessible to newcomers. For example, legend has it that there's [a Roslyn NuGet package](https://www.nuget.org/packages/Microsoft.Net.Compilers) that, once installed, automatically makes MSBuild use the version of `csc` contained in that package. Could we do something similar for EC#?

**Comment by qwertie** [comment]
It doesn't do anything magic, but I made a LeMP-tool package for @dadhi which, er, I think is invoked with `dotnet LeMP` but I've never actually used it!

Anyway it's finally done, you can run arbitrary code at compile time using the `compileTime` and `precompute` macros (among others). And the next release will have a `macro` macro for defining macros (it's based on `compileTime`, which in turn auto-registers macros when they have a `LexicalMacroAttribute` attached.)

The LeMP extension for Visual Studio used to be on the VS Marketplace, but I forgot to keep it updated so it got out of date. Now there's some kind of bureaucratic restriction where it won't let me upload a new version, so I deleted the old one instead. Hopefully I find a workaround soon (I normally publish the extension [here](https://github.com/qwertie/ecsharp/releases)).
